src/iceberg/expected.h

```diff
@@ -0,0 +1,2332 @@
+  move_assign_base() = default;
+  move_assign_base(const move_assign_base& rhs) = default;
+  move_assign_base(move_assign_base&& rhs) = default;
+  move_assign_base& operator=(const move_assign_base& rhs) = default;
+
+  constexpr move_assign_base& operator=(move_assign_base&& rhs)  //
+      noexcept(std::is_nothrow_move_constructible_v<E> &&
+               std::is_nothrow_move_assignable_v<E>) {
+    if (this->m_has_val && rhs.m_has_val) {
+      // no-op
+    } else if (this->m_has_val) {
+      std::construct_at(std::addressof(this->m_unexpect), std::move(rhs.m_unexpect));
+      this->m_has_val = false;
+    } else if (rhs.m_has_val) {
+      this->m_unexpect.~E();
+      this->m_has_val = true;
+    } else {
+      this->m_unexpect = std::move(rhs.m_unexpect);
+    }
+    return *this;
+  }
+};
+
+// This is needed to be able to construct the default_ctor_base which
+// follows, while still conditionally deleting the default constructor.
+struct default_constructor_tag {
+  explicit default_constructor_tag() = default;
+};
+
+// default_ctor_base will ensure that expected has a deleted default
+// consturctor if T is not default constructible.
+// This specialization is for when T is default constructible
+template <class T, class E, bool Enable = is_default_constructible_or_void_v<T>>
+struct default_ctor_base {
+  ~default_ctor_base() = default;
+  default_ctor_base() = default;
+  default_ctor_base(default_ctor_base const&) = default;
+  default_ctor_base(default_ctor_base&&) = default;
+  default_ctor_base& operator=(default_ctor_base const&) = default;
+  default_ctor_base& operator=(default_ctor_base&&) = default;
+
+  constexpr explicit default_ctor_base(default_constructor_tag) {}
+};
+
+// This specialization is for when T is not default constructible
+template <class T, class E>
+struct default_ctor_base<T, E, false> {
+  ~default_ctor_base() = default;
+  default_ctor_base() = delete;
+  default_ctor_base(default_ctor_base const&) = default;
+  default_ctor_base(default_ctor_base&&) = default;
+  default_ctor_base& operator=(default_ctor_base const&) = default;
+  default_ctor_base& operator=(default_ctor_base&&) = default;
+
+  constexpr explicit default_ctor_base(default_constructor_tag) {}
+};
+
+}  // namespace expected_detail
+
+/// An `expected<T, E>` object is an object that contains the storage for
+/// another object and manages the lifetime of this contained object `T`.
+/// Alternatively it could contain the storage for another unexpected object
+/// `E`. The contained object may not be initialized after the expected object
+/// has been initialized, and may not be destroyed before the expected object
+/// has been destroyed. The initialization state of the contained object is
+/// tracked by the expected object.
+
+template <class T, class E>
+class ICEBERG_EXPORT expected : private expected_detail::move_assign_base<T, E>,
+                                private expected_detail::default_ctor_base<T, E> {
```

Comment:
It was stated that we should not enforce `[[maybe_unused]]` via linter because we could apply it here. Should we do that?

Also, since stdlib doesn't declare it with that attribute, what will happen when we replace `iceberg::expected` with `std::expected`?

Reply:
> It was stated that we should not enforce `[[maybe_unused]]` via linter because we could apply it here. Should we do that?

You mean [[nodiscard]]? I changed with that.

> 
> Also, since stdlib doesn't declare it with that attribute, what will happen when we replace `iceberg::expected` with `std::expected`?

No idea, I hope our enforcement will encourage users to adopt a good code style.

Reply:
Oops, yes, `[[nodiscard]]`. 

But then that means we can't switch to std::expected down the line, because we've diverged.

Reply:
I'm not sure I understand this, IMO adding [[nodiscard]] to iceberg::expected will force users to handle the returned iceberg::expected, switch to std::expected will loosen this restriction. Please correct me if I'm wrong ;)

Reply:
Yes, but we don't want to loosen the restriction. New code will still be written after all, and suddenly users will lose an important compile time check.

Reply:
ok, do you think we should remove this class level [[nodiscard]] and add function level [[discard]]? I once had a idea adding a `Status` wrapper class around `iceberg::expected`, but I don't think that's a good idea.

Reply:
I think we have to use function level `[[nodiscard]]`, yes. We just won't have any lint to check it

Reply:
Is it possible to use the following code to wrap `std::expected`?
```
template <typename T, typename E>
struct [[nodiscard]] expected : std::expected<T, E> {
    using std::expected<T, E>::expected;
};
```

Reply:
I think it would technically work but I don't think subclassing STL classes is generally recommended (and I don't think expected is meant to be subclassed) so it's possible an implementation change or a different STL implementation would break us

Reply:
I agree that it is not a good practice to inherit `std::expected`. 

I'm not sure if this has been discussed already. We can configure `bugprone-unused-return-value` to check specific return type (including both the imported expected type and the std::expected from C++23): https://clang.llvm.org/extra/clang-tidy/checks/bugprone/unused-return-value.html#cmdoption-arg-CheckedReturnTypes.

Reply:
We can check it in our own codebase, but the purpose of the annotation is for downstream consumers, isn't it?

Reply:
I believe we've reached a consensus to add the [[nodiscard]] attribute to the exported interfaces. I keep the [[nodiscard]] attribute on the imported `expected` since I see no harm in leaving it here.